File: internal/integration/integration.go
// Package integration defines what every integration implements and what the
// framework gives it.
//
// Vocabulary: an Integration is the product (kubernetes, jira). A Connection
// is one configured system of that product (one cluster, one Jira site).
package integration

import (
	"context"
	"fmt"
	"log/slog"
	"net"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/roee-hersh/hallpass/internal/catalog"
	"github.com/roee-hersh/hallpass/internal/secret"
)

// Integration is the product.
type Integration interface {
	// Name is the value of the "integration" config key, lowercase.
	Name() string
	// Fields lists every config key this integration accepts besides the
	// common ones (id, integration, ca_file, tls_server_name, proxy_url, timeout).
	Fields() []Field
	// Actions lists what callers may ask about.
	Actions() []catalog.Action
	// New builds one connection from validated settings.
	New(ctx context.Context, s *Settings, d Deps) (Connection, error)
}

// Connection is one configured system.
type Connection interface {
	// ResolveIdentity maps the caller's user to the system's own account.
	// Return *Error with CodeUserNotFound or CodeUserAmbiguous when the
	// mapping fails; other errors mean the lookup itself failed.
	ResolveIdentity(ctx context.Context, u User) (Identity, error)
	// Check answers one question. It must return deny only when the system
	// positively says no. A returned error is turned into an unknown decision.
	Check(ctx context.Context, r CheckRequest) (Decision, error)
	// Probe verifies the credential and permissions without a user.
	Probe(ctx context.Context) (ProbeResult, error)
}

// ActionMatcher is optionally implemented by integrations whose action names
// are patterns (raw:<verb>:<resource>, app.action/<g>/<k>/<n>). It is consulted
// only when an exact match in Actions fails.
type ActionMatcher interface {
	MatchAction(name string) (catalog.Action, bool)
}

// User is what the caller sent.
type User struct {
	Email  string
	Groups []string
}

// Identity is the resolved account in the third-party system.
type Identity struct {
	// ID is the system's identifier (accountId, login, user id, username).
	ID string
	// Display is a short human-readable label for logs and reasons.
	Display string
	// Attrs carries small string facts (for example "is_admin"="true").
	Attrs map[string]string
	// Groups are group identifiers the system reported, if any.
	Groups []string
	// Native optionally keeps the integration's own decoded record.
	// It is cached, so it must be treated as read-only.
	Native any
}

// Attr returns one attribute or "".
func (i Identity) Attr(k string) string {
	if i.Attrs == nil {
		return ""
	}
	return i.Attrs[k]
}

// CheckRequest is one question for a Connection.
type CheckRequest struct {
	User     User
	Identity Identity
	Action   catalog.Action
	// ActionName is the caller's exact action string (matters for patterns).
	ActionName string
	Resource   catalog.Resource
}

// ProbeResult is what a successful probe reports.
type ProbeResult struct {
	// Summary is one line, for example "authenticated as bot@acme.com".
	Summary string
	// Warnings lists non-fatal findings such as an over-privileged credential.
	Warnings []string
}

// Field describes one config key an integration accepts.
type Field struct {
	Name     string
	Required bool
	// Secret fields must be env: or file: references.
	Secret bool
	// Ref names another integration; the value must be the id of a
	// connection of that integration. Used for keys ending in _connection.
	Ref string
	// Default applies when the key is absent.
	Default string
	// Enum restricts the accepted values.
	Enum []string
	// Description is one line for `hallpass catalog`.
	Description string
	// Validate optionally checks the value. It runs after Enum.
	Validate func(v string) error
}

// Common field constructors used by most integrations.

// URLField is the base URL of the system.
func URLField(required bool, desc string) Field {
	return Field{Name: "url", Required: required, Description: desc, Validate: ValidateHTTPSURL}
}

// CredentialField is the bot credential.
func CredentialField(required bool, desc string) Field {
	return Field{Name: "credential", Required: required, Secret: true, Description: desc}
}

// ConnectionRefField references a connection of another integration.
func ConnectionRefField(name, integration string, required bool, desc string) Field {
	return Field{Name: name, Ref: integration, Required: required, Description: desc}
}

var fieldNameRe = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

// ValidateFields checks an integration's field declarations at registration.
func ValidateFields(fields []Field) error {
	seen := map[string]bool{}
	for _, f := range fields {
		if !fieldNameRe.MatchString(f.Name) {
			return fmt.Errorf("field %q: name must match %s", f.Name, fieldNameRe)
		}
		if seen[f.Name] {
			return fmt.Errorf("field %q declared twice", f.Name)
		}
		seen[f.Name] = true
		if CommonFields[f.Name] {
			return fmt.Errorf("field %q is a common field and may not be redeclared", f.Name)
		}
		if f.Ref != "" && !strings.HasSuffix(f.Name, "_connection") {
			return fmt.Errorf("field %q references a connection so its name must end in _connection", f.Name)
		}
		if f.Secret && f.Default != "" {
			return fmt.Errorf("field %q: secret fields cannot have defaults", f.Name)
		}
	}
	return nil
}

// CommonFields are accepted on every connection and handled by the framework.
var CommonFields = map[string]bool{
	"id": true, "integration": true,
	"ca_file": true, "tls_server_name": true, "proxy_url": true, "timeout": true,
}

// ValidateHTTPSURL accepts https:// URLs with no query, fragment or userinfo.
// Plain http:// is allowed only when the host is exactly "localhost" or a
// loopback IP address (127.0.0.0/8, ::1), for local testing; a name that
// merely starts with "localhost" or "127.0.0.1" resolves wherever DNS says
// and would carry the credential in clear text, so it is rejected.
func ValidateHTTPSURL(v string) error {
	if v == "" {
		return nil
	}
	if strings.ContainsAny(v, " \t\r\n#?") {
		return fmt.Errorf("url %q must not contain whitespace, '?' or '#'", v)
	}
	u, err := url.Parse(v)
	if err != nil {
		return fmt.Errorf("url %q: %v", v, err)
	}
	if u.User != nil {
		return fmt.Errorf("url %q must not contain userinfo", v)
	}
	switch u.Scheme {
	case "https":
		return nil
	case "http":
		if isLoopbackHost(u.Hostname()) {
			return nil
		}
	}
	return fmt.Errorf("url %q must start with https://", v)
}

// isLoopbackHost reports whether host is "localhost" or a loopback IP.
func isLoopbackHost(host string) bool {
	if host == "localhost" {
		return true
	}
	ip := net.ParseIP(host)
	return ip != nil && ip.IsLoopback()
}

// Settings are the validated config values of one connection.
type Settings struct {
	ID          string
	Integration string

	// Common transport options.
	CAFile        string
	TLSServerName string
	ProxyURL      string
	Timeout       time.Duration

	values  map[string]string
	secrets map[string]secret.Secret
}

// NewSettings builds Settings. Tests and the config loader use it.
func NewSettings(id, integration string, values map[string]string, secrets map[string]secret.Secret) *Settings {
	s := &Settings{ID: id, Integration: integration, values: map[string]string{}, secrets: map[string]secret.Secret{}}
	for k, v := range values {
		s.values[k] = v
	}
	for k, v := range secrets {
		s.secrets[k] = v
	}
	return s
}

// Get returns a non-secret value (defaults already applied) or "".
func (s *Settings) Get(key string) string { return s.values[key] }

// Has reports whether a non-secret key was set (or defaulted).
func (s *Settings) Has(key string) bool { _, ok := s.values[key]; return ok }

// Bool parses a "true"/"false" value; absent means def.
func (s *Settings) Bool(key string, def bool) bool {
	switch strings.ToLower(s.values[key]) {
	case "true", "yes", "1":
		return true
	case "false", "no", "0":
		return false
	default:
		return def
	}
}

// Secret returns a secret reference; the zero Secret when absent.
func (s *Settings) Secret(key string) secret.Secret { return s.secrets[key] }

// Keys lists the set non-secret keys, sorted.
func (s *Settings) Keys() []string {
	ks := make([]string, 0, len(s.values))
	for k := range s.values {
		ks = append(ks, k)
	}
	sort.Strings(ks)
	return ks
}

// Deps is what the framework hands to Integration.New.
type Deps struct {
	Logger *slog.Logger
	// Connection returns another built connection by id. Only ids named in a
	// Ref field of this integration are resolvable.
	Connection func(id string) (Connection, error)
	// HTTPClient builds a client from the connection's transport settings.
	HTTPClient func(s *Settings) (*http.Client, error)
	// Now is the clock. Tests replace it.
	Now func() time.Time
}

// DefaultTimeout applies when a connection sets none.
const DefaultTimeout = 8 * time.Second

// EffectiveTimeout returns the connection's timeout or the default.
func (s *Settings) EffectiveTimeout() time.Duration {
	if s.Timeout > 0 {
		return s.Timeout
	}
	return DefaultTimeout
}

File: hallpass-py/src/hallpass/core/integration.py
"""What every integration implements and what the framework gives it.

Vocabulary: an Integration is the product (kubernetes, jira). A Connection
is one configured system of that product (one cluster, one Jira site).
"""

from __future__ import annotations

import ipaddress
import re
import threading
import time
import urllib.parse
from abc import ABC, abstractmethod
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from hallpass.core.catalog import Action, Resource, validate_action_name
from hallpass.core.context import Context
from hallpass.core.decision import Decision
from hallpass.core.log import Logger
from hallpass.core.secret import Secret

if TYPE_CHECKING:
    from hallpass.net.httpx import Transport

__all__ = [
    "COMMON_FIELDS",
    "DEFAULT_TIMEOUT",
    "CheckRequest",
    "Connection",
    "Deps",
    "Field",
    "Identity",
    "Integration",
    "ProbeResult",
    "Registry",
    "Settings",
    "User",
    "connection_ref_field",
    "credential_field",
    "find_action",
    "is_loopback_host",
    "url_field",
    "validate_fields",
    "validate_https_url",
]

# The per-check budget when a connection sets none.
DEFAULT_TIMEOUT = 8.0


@dataclass(frozen=True)
class User:
    """What the caller sent."""

    email: str
    groups: tuple[str, ...] = ()


@dataclass(frozen=True)
class Identity:
    """The resolved account in the third-party system."""

    # The system's identifier (accountId, login, user id, username).
    id: str = ""
    # A short human-readable label for logs and reasons.
    display: str = ""
    # Small string facts (for example "is_admin" -> "true").
    attrs: Mapping[str, str] = field(default_factory=dict)
    # Group identifiers the system reported, if any.
    groups: tuple[str, ...] = ()
    # Optionally the integration's own decoded record. It is cached, so it
    # must be treated as read-only.
    native: Any = field(default=None, compare=False)

    def attr(self, k: str) -> str:
        return self.attrs.get(k, "") if self.attrs else ""


@dataclass(frozen=True)
class CheckRequest:
    """One question for a Connection."""

    user: User
    identity: Identity
    action: Action
    # The caller's exact action string (matters for patterns).
    action_name: str
    resource: Resource


@dataclass(frozen=True)
class ProbeResult:
    # One line, for example "authenticated as bot@acme.com".
    summary: str = ""
    # Non-fatal findings such as an over-privileged credential.
    warnings: tuple[str, ...] = ()


@dataclass(frozen=True)
class Field:
    """One config key an integration accepts."""

    name: str
    required: bool = False
    # Secret fields must be env: or file: references.
    secret: bool = False
    # Names another integration; the value must be the id of a connection
    # of that integration. Used for keys ending in _connection.
    ref: str = ""
    # Applies when the key is absent.
    default: str = ""
    # Restricts the accepted values.
    enum: tuple[str, ...] = ()
    # One line for `hallpass catalog`.
    description: str = ""
    # Optionally checks the value (raise ValueError). Runs after enum.
    validate: Callable[[str], None] | None = field(default=None, compare=False)


def url_field(required: bool, desc: str) -> Field:
    """The base URL of the system."""
    return Field(name="url", required=required, description=desc, validate=validate_https_url)


def credential_field(required: bool, desc: str) -> Field:
    """The bot credential."""
    return Field(name="credential", required=required, secret=True, description=desc)


def connection_ref_field(name: str, integration: str, required: bool, desc: str) -> Field:
    """A reference to a connection of another integration."""
    return Field(name=name, ref=integration, required=required, description=desc)


_FIELD_NAME_RE = re.compile(r"[a-z][a-z0-9_]*")

# Accepted on every connection and handled by the framework.
COMMON_FIELDS = frozenset({"id", "integration", "ca_file", "tls_server_name", "proxy_url", "timeout"})


def validate_fields(fields: list[Field] | tuple[Field, ...]) -> None:
    """Check an integration's field declarations at registration."""
    seen: set[str] = set()
    for f in fields:
        if not _FIELD_NAME_RE.fullmatch(f.name):
            raise ValueError(f'field "{f.name}": name must match ^[a-z][a-z0-9_]*$')
        if f.name in seen:
            raise ValueError(f'field "{f.name}" declared twice')
        seen.add(f.name)
        if f.name in COMMON_FIELDS:
            raise ValueError(f'field "{f.name}" is a common field and may not be redeclared')
        if f.ref and not f.name.endswith("_connection"):
            raise ValueError(f'field "{f.name}" references a connection so its name must end in _connection')
        if f.secret and f.default:
            raise ValueError(f'field "{f.name}": secret fields cannot have defaults')


def validate_https_url(v: str) -> None:
    """Accept https:// URLs with no query, fragment or userinfo.

    Plain http:// is allowed only when the host is exactly "localhost" or a
    loopback IP address, for local testing; a name that merely starts with
    "localhost" or "127.0.0.1" resolves wherever DNS says and would carry
    the credential in clear text, so it is rejected.
    """
    if v == "":
        return
    if any(c in v for c in " \t\r\n#?"):
        raise ValueError(f"url \"{v}\" must not contain whitespace, '?' or '#'")
    try:
        u = urllib.parse.urlsplit(v)
        host = u.hostname or ""
        _ = u.port  # a malformed port raises
    except ValueError as e:
        raise ValueError(f'url "{v}": {e}') from None
    if "@" in u.netloc:
        raise ValueError(f'url "{v}" must not contain userinfo')
    if u.scheme == "https":
        return
    if u.scheme == "http" and is_loopback_host(host):
        return
    raise ValueError(f'url "{v}" must start with https://')


def is_loopback_host(host: str) -> bool:
    """host is "localhost" or a loopback IP address."""
    if host == "localhost":
        return True
    try:
        return ipaddress.ip_address(host.strip("[]")).is_loopback
    except ValueError:
        return False


class Settings:
    """The validated config values of one connection."""

    def __init__(
        self,
        id: str,
        integration: str,
        values: Mapping[str, str] | None = None,
        secrets: Mapping[str, Secret] | None = None,
        *,
        ca_file: str = "",
        tls_server_name: str = "",
        proxy_url: str = "",
        timeout: float = 0.0,
    ) -> None:
        self.id = id
        self.integration = integration
        self.ca_file = ca_file
        self.tls_server_name = tls_server_name
        self.proxy_url = proxy_url
        self.timeout = timeout
        self._values: dict[str, str] = dict(values or {})
        self._secrets: dict[str, Secret] = dict(secrets or {})

    def get(self, key: str) -> str:
        """A non-secret value (defaults already applied) or ""."""
        return self._values.get(key, "")

    def has(self, key: str) -> bool:
        """Whether a non-secret key was set (or defaulted)."""
        return key in self._values

    def bool(self, key: str, default: bool) -> bool:
        v = self._values.get(key, "").lower()
        if v in ("true", "yes", "1"):
            return True
        if v in ("false", "no", "0"):
            return False
        return default

    def secret(self, key: str) -> Secret:
        """A secret reference; the empty Secret when absent."""
        return self._secrets.get(key, Secret())

    def keys(self) -> list[str]:
        return sorted(self._values)

    def effective_timeout(self) -> float:
        return self.timeout if self.timeout > 0 else DEFAULT_TIMEOUT

    def __repr__(self) -> str:
        return f"Settings(id={self.id!r}, integration={self.integration!r})"


@dataclass
class Deps:
    """What the framework hands to Integration.new."""

    logger: Logger
    # Another built connection by id. Only ids named in a ref field of this
    # integration are resolvable.
    connection: Callable[[str], Connection]
    # A transport built from the connection's transport settings.
    http_client: Callable[[Settings], Transport]
    # The clock (epoch seconds). Tests replace it.
    now: Callable[[], float] = time.time


class Connection(ABC):
    """One configured system."""

    @abstractmethod
    def resolve_identity(self, ctx: Context, u: User) -> Identity:
        """Map the caller's user to the system's own account.

        Raise HallpassError with USER_NOT_FOUND or USER_AMBIGUOUS when the
        mapping fails; any other exception means the lookup itself failed.
        """

    @abstractmethod
    def check(self, ctx: Context, r: CheckRequest) -> Decision:
        """Answer one question. Return deny only when the system positively
        says no. A raised exception is turned into an unknown decision."""

    @abstractmethod
    def probe(self, ctx: Context) -> ProbeResult:
        """Verify the credential and permissions without a user."""


class Integration(ABC):
    """The product."""

    @abstractmethod
    def name(self) -> str:
        """The value of the "integration" config key, lowercase."""

    @abstractmethod
    def fields(self) -> list[Field]:
        """Every config key this integration accepts besides the common ones."""

    @abstractmethod
    def actions(self) -> list[Action]:
        """What callers may ask about."""

    @abstractmethod
    def new(self, ctx: Context, s: Settings, d: Deps) -> Connection:
        """Build one connection from validated settings."""

    def match_action(self, name: str) -> Action | None:
        """For integrations whose action names are patterns; consulted only
        when an exact match in actions() fails."""
        return None


def find_action(i: Integration, name: str) -> Action | None:
    """Exact match in actions() first, then match_action."""
    for a in i.actions():
        if not a.pattern and a.name == name:
            return a
    return i.match_action(name)


_NAME_RE = re.compile(r"[a-z][a-z0-9]*")


class Registry:
    """The known integrations by name."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._items: dict[str, Integration] = {}

    def register(self, i: Integration) -> None:
        """Add an integration after validating its name, fields and actions.
        A programming error raises: registration happens at import time."""
        name = i.name()
        if not _NAME_RE.fullmatch(name):
            raise ValueError(f'integration name "{name}" must match ^[a-z][a-z0-9]*$')
        try:
            validate_fields(i.fields())
        except ValueError as e:
            raise ValueError(f"integration {name}: {e}") from None
        seen: set[str] = set()
        for a in i.actions():
            if not a.pattern:
                try:
                    validate_action_name(a.name)
                except ValueError as e:
                    raise ValueError(f'integration {name}: action "{a.name}": {e}') from None
            if a.name in seen:
                raise ValueError(f'integration {name}: action "{a.name}" declared twice')
            seen.add(a.name)
        with self._lock:
            if name in self._items:
                raise ValueError(f"integration {name} registered twice")
            self._items[name] = i

    def lookup(self, name: str) -> Integration | None:
        with self._lock:
            return self._items.get(name)

    def names(self) -> list[str]:
        with self._lock:
            return sorted(self._items)
